Runestone to PreTeXt

bnmnetp opened the pull request:
This is a draft PR, a work in progress to create a second output mode for Runestone books.  Instead of generating the usual html this PR aims to output XML that can be further converted into PreTeXt (PTX).  After conversion the source can be built with the PreTeXt tools into html, epub, braille pdf, etc.   

Why?

1. The goal is for the Runestone project to let go of the authoring side of books and leave that to the PreTeXt project.  A very nice division of labor. That allows us to focus on the components and the server
2. PreTeXt books work on the Runestone Academy platform and will get better over time.  But only if there are books, in PTX that use all of the features!  Its a push-pull relationship
3. To move forward to more fully customizing textbooks requires a more structured authoring language. PTX is already there.
4. If people really want to keep writing books in restructuredText they can... but this also provides a path to other output formats or the eventual move to PTX.

The main challenges of this PR are to get decent looking XML output for our directives.  I'm targeting thinkcspy as it is the oldest in the family, and could use some love.

Unfortunately generating XML has uncovered a few problems with how we coded the HTML generation.

I would welcome early comments on the changes I have made. @bjones1 I am looking in your direction... I would also welcome any help in the "wood chopping" phase of this conversion.  The basic flaw in our implementation is that we did not follow the (undocumented) interface for nodes derived from docutils `nodes.General` and `nodes.Element` that is we should not have a constructor that takes our own parameters and we should not add arbitrary attributes to a derived node.  Instead we should treat a node like a dictionary and add new values to the node that way.

bjones1 commented on runestone/mchoice/multiplechoice.py:
I'd suggest adding more text to this comment. Perhaps:

In the following text, let `ro = node["runestone_options"]`. 

For an answer with the label `foo`, `ro` contains the following key/value pairs:

- `answer_foo`: the answer text
- `feedback_foo`: feedback text. This key/value pair is optional.

In addition, `ro["correct"]` == foo if this is the correct answer.

The following code unpacks this convoluted data structure into more sane values of:
- `ro["alabel"]` = foo
- `ro["atext"]` = the answer text or ""
- `ro["is_correct"]` = "true" or "false"

TODO: do this unpack/rearrange earlier in the code...

bnmnetp replied:
I hear you, but global search and replace with regex saves me a LOT of typing.

bjones1 commented on runestone/mchoice/multiplechoice.py:
```diff
@@ -210,12 +249,15 @@ def run(self):
+        mcNode["runestone_options"] = self.options
+        mcNode["source"], mcNode["line"] = self.state_machine.get_source_and_line(
+            self.lineno)
+        mcNode["template_start"] = TEMPLATE_START
```
Why assign constants to each instance of `mcNode`? This would make sense as globals in this module instead.

bjones1 commented on scripts/dist2xml.py:
```diff
@@ -0,0 +1,9 @@
+import json
```
It seems like we have a BUNCH of scripts these days, and I know know what most of them do, or even if they're currently used. I would suggest:

1. Create a `contents.rst` file which gives a one-sentence description of each scritp.
2. Eventually, port scripts to Python. Bash scripts are a pain to maintain; the `xqt` function in `ci_utils` makes it easy to run programs in a Python script with bash-like ease.

bnmnetp replied:
I find bash scripts to be quite easy to write, read,  and maintain.

Most of them are for my use in maintaining and running Academy

bjones1 commented on scripts/makePtx.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Noooooo....more bash scripts..... :) With no docs/explanations.... :(

bnmnetp commented on runestone/activecode/activecode.py:
```diff
@@ -79,16 +80,18 @@ def setup(app):
+
+def visit_ac_xml(self, node):
+    res = TEMPLATE_START % node["runestone_options"]
+    self.output.append(res)
```
Because the xml writer uses `output` instead of `body`.  Nice huh??

bjones1 replied:
docutils, ever obscure and unpredictable....sigh...

bnmnetp commented on runestone/mchoice/multiplechoice.py:
Thats fine, the main goal here is to get things back to a state where we can build both html and xml.

If you have suggestions for how we might refactor the different node functions to make them as DRY as possible that is great, but rewriting working code to simply use a different style is not something I'm too interested in doing.

bjones1 replied:
Sounds good. I'll provide comments more in the DRY direction, then. Getting them as DRY as possible, given we'll need to repeat this basic structure a bunch of times, is definitely important.

bnmnetp commented on runestone/activecode/activecode.py:
```diff
@@ -448,8 +451,10 @@ def run(self):
 
-        acnode = ActivcodeNode(self.options, rawsource=self.block_text)
-        acnode.source, acnode.line = self.state_machine.get_source_and_line(self.lineno)
+        acnode = ActivecodeNode(self.options, rawsource=self.block_text)
```
I experimented with different ways of trying to do something like that but was unsuccessful.

bjones1 replied:
Docutils can be so obscure at times -- sigh...

bnmnetp commented on runestone/mchoice/multiplechoice.py:
Take a look at what I just pushed.  I think this will work well for refactoring.

bjones1 replied:
I like it! That's really nice -- I agree, use this for the refactor / including XML output.

bnmnetp commented:
@bjones1 - I've updated the syntax for all of the nodes.  I've only added xml for a couple so far.  All unit tests are passing locally and on github...


## Comments/Decisions

1. I am using Sphinx 4.4.0 and docutils 0.17.1 locally.  On github the tests are still using docutils 0.16 -- read on for more
2. There are errors in the unit tests from 4.4.0 on master relating to things that used to be warnings now being reported as errors by Sphinx.  I have fixed those tests on this branch, so if you could pay special attention to those test changes that would be great.
1. In order to get the XML output we want we need at least docutils 0.17.1
3. docutils 0.17.1 breaks our styling and makes the text on our pages full width and quite ugly.   So we will want to keep 0.16 for production.  @ascholerChemeketa looked at what would need to be done to update the css and thought it was a largish project.
4. Since all of the tests are passing for the html documents with docutils 0.16 I suggest we merge that to make a cleaner break.  We may need to keep the xml conversion stuff on its own branch for a while, but these changes are pretty extensive and I don't want to end up managing tons of conflicts.
5. docutils 0.18.1 is out and may fix the formatting, but I'm not sure and Sphinx has not
6. I don't think you need to review every line many many lines were changed with global search/replace so I have not looked at the code to make improvements in any other way and I'm not interested in polishing all of this right now.

bnmnetp commented:
Update:  I think there is an easy change that fixes *almost* all the problems with `docutils 0.17.1` The major difference is a `<div class="section">` getting changed to use semantic tags (just `<section>`) modifying `runestone-custom-sphinx-bootstrap.css` to reflect that change seems to restore a proper look.

bjones1 commented:
I'm in favor of merging this soon-ish, especially if there's a good path to get CSS happy again!

bjones1 commented on runestone/common/question_number.py:
```diff
@@ -7,7 +7,7 @@
 # Third-party
 from docutils import nodes
-
+import pdb
```
Remove before merge.

bjones1 commented on runestone/mchoice/multiplechoice.py:
```diff
@@ -13,6 +13,7 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
+import pdb
```
Remove before merge

bjones1 commented:
The changes in tests seem reasonable to me.

bnmnetp commented:
I can merge this as this part of it works perfectly well with docutils 0.16.1.   Andrew is working on some updates that we should have soon that will allow us to go to 0.17.1 and have nice looking components which will have much more flexibility in terms of adjusting widths.

bjones1 commented:
That sounds good to me!
